### ISSUE TITLE
Regression in m2-oss/calypso - rhs of a given with type bounded param unable to resolve certain other givens

### ISSUE THREAD

**Issue by jchyb** [lampepfl/dotty#19407]
## Compiler version

`3.4.0-RC1-bin-20240104-2746ee8-NIGHTLY` (last successful build `3.4.0-RC1-bin-20231207-16f1680-NIGHTLY`), via the community build: https://github.com/VirtusLab/community-build3/actions/runs/7427597656/job/20238015937

## Minimized code

```Scala
trait GeneratedEnum
trait Decoder[A]

object Decoder:
  given Decoder[Int] = ???

object GeneratedEnumDecoder:

  given [A <: GeneratedEnum]: Decoder[A] =
    summon[Decoder[Int]]
    ???

```

## Output

```scala
[error] -- [E172] Type Error: /Users/jchyb/Documents/workspace/calypso/modules/scalapb/src/main/scala/ru/m2/calypso/scalapb/GeneratedEnumDecoder.scala:10:24 
[error] 10 |    summon[Decoder[Int]]
[error]    |                        ^
[error]    |No given instance of type Decoder[Int] was found for parameter x of method summon in object Predef.
[error]    |I found:
[error]    |
[error]    |    GeneratedEnumDecoder.given_Decoder_A[A]
[error]    |
[error]    |But given instance given_Decoder_A in object GeneratedEnumDecoder does not match type Decoder[Int].
[error] one error found
```

## Expectation
no errors

**Comment by odersky** [comment]
What happens here is that the given in DeneratedDecoder itself is picked as the resolved instance. The whole is solved with #19392 but one has to compile with `-source future`.

It seems what happens is that previously the given in `GeneratedEnumDecoder` was rejected since `Int` does not conform to `GeneratedEnum`. But maybe now the bounds are not checked eagerly (I am only guessing). Can we get a bisect that shows where things go wrong?

**Comment by odersky** [comment]
It's actually #19392 that broke it. I.e. it now works with -source future, but not now. I'll look into it.
